### PR TITLE
#1125 Fix parsing of unusual trailing expressions separated by whitespace

### PR DESCRIPTION
**Pull request by technohippy**
Concerning:

- Is "unexpected token error" appropriate for this error?
- Serialized source text still has ";". Is it OK?

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/tokenizer.ts [hunk 1]
```diff
@@ -1697,6 +1697,18 @@ export class Tokenizer extends DiagnosticEmitter {
     return this.readHexadecimalEscape(4, startIfTaggedTemplate);
   }
 
+  hasNoMoreExpressionInLine(): bool {
+    let pos = this.pos;
+    const end = this.end;
+    do {
+      const c = this.source.text.charCodeAt(pos);
```
Comment (dcodeIO) [inline src/tokenizer.ts:1704]: Looks like `pos` can be `== end` here, so the access can fail. A `while` instead would probably work.

File: src/tokenizer.ts [hunk 2]
```diff
@@ -1697,6 +1697,18 @@ export class Tokenizer extends DiagnosticEmitter {
     return this.readHexadecimalEscape(4, startIfTaggedTemplate);
   }
 
+  hasNoMoreExpressionInLine(): bool {
```
Comment (dcodeIO) [inline src/tokenizer.ts:1700]: Technically not looking for an expression, but for non-whitespace. Not sure what would be a good name, hmm.

**Comment by dcodeIO** [comment]
I wonder whether this is only applicable to `var`iables and `return`s. Another one that comes to mind is `throw`. Do you know what the spec says exactly for this case? If it runs deeper, perhaps a fix could be more general.

**Comment by technohippy** [comment]
> I wonder whether this is only applicable to variables and returns. Another one that comes to mind is throw. Do you know what the spec says exactly for this case? If it runs deeper, perhaps a fix could be more general.

I agree. I don't know the spec well but I should consider more deeply.

Anyway, this change counts some valid codes like below as an error.

```js
function foo(): i32 {
  return 123 }
```

I'll close this pullreq for now and reopen it after making more proper solution.

**Comment by technohippy** [comment]
I read the spec but it's too complecated for me to handle all cases.

- https://tc39.es/ecma262/#sec-automatic-semicolon-insertion

Anyway, this pull req can fix issue #1125. If other cases are found, please replace `tn.skip(Token.SEMICOLON)` to `this.checkRuleCompleted(tn)`.

File: src/tokenizer.ts (hunk 1 above)
Comment (technohippy) [reply src/tokenizer.ts:1703]: fixed

**Comment by dcodeIO** [comment]
Thanks :) There are likely more ASI-related issues, but seems good to have this merged for the (probably most prominent) cases covered herein.

**Pull request merged**